util/dvsim/SGE.py

```diff
@@ -0,0 +1,345 @@
+
+            try:
+                qliststr = _exec(self.cmd_qconf + ' -sql')
+            except IOError:
+                error_msg = 'Error querying queue configuration'
+                logger.error(error_msg)
+                raise IOError(error_msg)
+
+            self.q = qliststr.replace('\n', ',')[:-1]
+
+            logger.info("""Sun Grid Engine handler initialized
+Queues detected: %s""", self.q)
+
+        else:
+            self.q = q
+
+        self.cmd_qsub = os.path.join(path, 'qsub')
+        self.cmd_qstat = os.path.join(path, 'qstat')
+
+    def wait(self, jobid, interval=10, name=None, pbar=None,
+             pbar_mode=None):
+        """Waits for job running on SGE Grid Engine environment to finish.
+
+        If you are just waiting for one job, this becomes a dumb substitute for
+        the -sync y option which can be specified to qsub.
+
+        Inputs:
+
+        jobid
+        interval - Polling interval of SGE queue, in seconds. (Default: 10)
+        """
+
+        logger = logging.getLogger('SGE.wait')
+
+        dowait = True
+        while dowait:
+            p = subprocess.Popen(self.cmd_qstat, shell=True,
+                                 stdout=subprocess.PIPE)
+            pout, _ = p.communicate()
+
+            if pbar is not None:
+                logger.error('Progress bar handling not implemented')
+                # if pbar_mode == 'quantum':
+                #    pbar.update(getstarfleet(getouts()))
+                # elif pbar_mode == 'mdrun':
+                #    pbar.update(getstep())
+
+            dowait = False
+            for line in pout.split('\n'):
+                t = line.split()
+                if len(t) >= 5:  # Find a line with useful info
+                    if t[0] == str(jobid) and re.search(t[4], 'qwrt'):
+                        # Job must be queued, running or being transferred
+                        dowait = True
+                        break
+                    if t[0] == str(jobid) and re.search(t[4], 'acuE'):
+                        # Job or host in error state
+                        logger.warning('Job %d in error state', str(jobid))
+
+            if dowait:
+                time.sleep(interval)
+                if name is None:
+                    logger.info("Time %s: waiting for jobid %s to finish",
+                                time.ctime(), str(jobid))
+                else:
+                    logger.info("Time %s: waiting for job '%s' (jobid %s) to \
+finish", time.ctime(), name, str(jobid))
+
+    def submit(self, job, array=False, useenvironment=True, usecwd=True,
+               name=None, stdin=None, stdout=None, stderr=None,
+               joinstdouterr=True, nproc=1, wait=True, lammpi=True):
+        """Submits a job to SGE
+
+        Returns jobid as a number"""
+
+        logger = logging.getLogger('SGE.submit')
+
+        logger.info("Submitting job:    \x1b[1;91m%-50s\x1b[0m Stdout: %s \
```

Comment:
This is a bizarre message... Perhaps this is some curses sequence. Could you add some comment describing what it is meant to do?

Reply:
fixed